File: heaps/find_median_from_data_stream.java
class MedianFinder
{
    private PriorityQueue<Integer> minHeap = new PriorityQueue<>();
    private PriorityQueue<Integer> maxHeap =
        new PriorityQueue<>(Comparator.reverseOrder());
    
    public void addNum(int num)
    {
        maxHeap.add(num);
        minHeap.add(maxHeap.remove());

        if (minHeap.size() > maxHeap.size())
        {
            maxHeap.add(minHeap.remove());
        }
    }

    public double findMedian()
    {
        if (maxHeap.size() > minHeap.size())
        {
            return maxHeap.peek();
        }

        return (minHeap.peek() + maxHeap.peek()) / 2.0;
    }
}

File: heaps/find_median_from_data_stream.py
from heapq import *

class MedianFinder:
    def __init__(self):
        self.min_heap = []
        self.max_heap = []
    
    def add_num(self, num: int) -> None:
        heappush(self.max_heap, -num)
        heappush(self.min_heap, -heappop(self.max_heap))

        if len(self.min_heap) > len(self.max_heap):
            heappush(self.max_heap, -heappop(self.min_heap))
    
    def find_median(self) -> float:
        if len(self.max_heap) > len(self.min_heap):
            return -self.max_heap[0]
        
        return (self.min_heap[0] - self.max_heap[0]) / 2.0
